### ISSUE TITLE
It's possible to commit a write transaction after becoming read-only with mvcc

### ISSUE THREAD

**Issue by sergepetrenko** [tarantool/tarantool#9937]
**Bug description**

Tarantool version: 3.1.0-entrypoint-234-g4466deafee

Our read-only checks only work when corresponding write statements are executed, and not on transaction commit.
So if code executing a transaction writes something and then yields before commit (which's possible with MVCC), and the node becomes read-only during this yield, the commit will happen anyway, writing data on a read-only node.

Since the WAL write is initiated only on transaction commit, this means a node may continue writing data after becoming read-only, breaking all the expectations about being read-only.

**Steps to reproduce**

Run a single tarantool instance in interactive mode and issue:

```lua
-- Step 1.
box.cfg{memtx_use_mvcc_engine = true}
s = box.schema.space.create('test')
_ = s:create_index('pk')

-- Step 2.
fiber = require('fiber')
log = require('log')
fiber.new(function()
    box.begin()
    log.info("read-only status before yield is %s", box.info.ro)
    s:insert{1}
    fiber.sleep(5)
    log.info("read-only status after yield is %s", box.info.ro)
    box.commit()
end)
box.cfg{read_only = true}

-- Step 3. Wait a bit (until the transaction above is committed), then do:
s:fselect{}
```
**Actual behavior**

During step 2 the logs will show:
```
2024-04-11 13:47:18.040 [30674] main/117/lua/tarantool I> read-only status before yield is false
tarantool> box.cfg{read_only = true}
2024-04-11 13:47:18.345 [30674] main/104/interactive/box.load_cfg I> set 'read_only' configuration option to true
2024-04-11 13:47:23.038 [30674] main/117/lua/tarantool I> read-only status after yield is true
```
And step 3 output:
```

tarantool> s:fselect{}
---
- |-
  +-----+
  |col1 |
  +-----+
  |  1  |
  +-----+
...
```

**Expected behavior**

`box.commit()` from the example above should fail and throw an error.
Something like: `The node became read-only while the transaction was processed`.

**Comment by sergepetrenko** [comment]
> @sergepetrenko Don't we have the same situation without mvcc, like described in https://github.com/tarantool/tarantool/issues/10142? When we're waiting for disk write to happen, read-only can become true and we'll successfully commit transaction after becoming read only node.

@Serpentian yes, but the difference in your example is that the node transitions to ro after `box.commit()` is called. I don't see why this is a problem. And it seems impossible to catch and fail such commit() calls. Because the data might be already written.

If you're thinking about some failover logic, where we have to wait for the node to stop writing anything new before we appoint another  leader, then that's what #10142 is about.

And together these two tickets accomplish the task quite well. Without this task we would never be sure that instance vclock is really stabilized after a `wal_sync()`, because some yielding mvcc transaction could wake up anytime **after** the transition to ro and after `wal_sync()`, and write something extra.

**Comment by Serpentian** [comment]
The quesiton, which bothers me, whether we need `wal_sync`. 

Am I right, that we need `wal_sync` in https://github.com/tarantool/tarantool/issues/10142, because we may start writing to the disk in `wal_write_to_disk` (and consequently we passed read-only check), but as wal didn't end writing, we may get irrelevant `box.info.lsn` after switching to read-only mode, which will be actualized only after `wal_sync`?

And another question. Why don't we want implicitly sync wal before switching to `ro`. Why do we introduce another hand for users? We have so many of them(

**Comment by sergepetrenko** [comment]
> Am I right, that we need wal_sync in https://github.com/tarantool/tarantool/issues/10142, because we may start writing to the disk in wal_write_to_disk (and consequently we passed read-only check), but as wal didn't end writing, we may get irrelevant box.info.lsn after switching to read-only mode, which will be actualized only after wal_sync?

Yes, exactly. This ticket makes sure that a node doesn't submit new WAL tasks after transition to `ro`, and the ticket about `wal_sync` lets the user wait until all the tasks in progress are finished.

> And another question. Why don't we want implicitly sync wal before switching to ro. Why do we introduce another hand for users? We have so many of them(

This would make a previously non-yielding configuration change yield. I think some users might rely on the fact that it doesn't yield. Besides, not every application needs a wal_sync after ro transition.

**Comment by sergepetrenko** [comment]
See also https://github.com/tarantool/tarantool/issues/4016

**Comment by sergepetrenko** [comment]
Backport to 2.11-3.3 is non-trivial, so we'll leave this fix only in 3.4 and master.
